src/compiler/parser.ts

```diff
@@ -2649,6 +2694,26 @@ namespace ts {
             return token() === SyntaxKind.CloseParenToken || isStartOfParameter() || isStartOfType();
         }
 
+        function parseJSDocPostfixTypeOrHigher(): TypeNode {

```

Comment:
Move this next to (or into) `parseTypeOperatorOrHigher`.
Can we parse `foo?[]` or `foo![]`? Could merge with `parseArrayTypeOrHigher` (also only called in one place).
I took a stab at making this neater:
```ts
function parseJSDocPostfixTypeOrHigher(): TypeNode {
    const type = parseArrayTypeOrHigher();
    const kind = getKind(token());
    if (!kind) return type;

    const postfix = createNode(kind, type.pos) as JSDocOptionalType | JSDocNonNullableType | JSDocNullableType;
    postfix.type = type;
    return finishNode(postfix);

    function getKind(tokenKind: SyntaxKind): SyntaxKind | undefined {
        switch (tokenKind) {
            case SyntaxKind.EqualsToken:
                // only parse postfix = inside jsdoc, because it's ambiguous elsewhere
                return contextFlags & NodeFlags.JSDoc ? SyntaxKind.JSDocOptionalType : undefined;
            case SyntaxKind.ExclamationToken:
                return SyntaxKind.JSDocNonNullableType;
            case SyntaxKind.QuestionToken:
                return SyntaxKind.JSDocNullableType;
        }
    }
}
```